lib/commons-core/src/main/java/org/apache/olingo/commons/core/edm/primitivetype/EdmDateTimeOffset.java

```diff
@@ -33,245 +38,142 @@
-      for (int d = 100 * (isNano ? 1000 * 1000 : 1); d > 0; d /= 10) {
-        final byte digit = (byte) (fractionalSeconds % (d * 10) / d);
-        if (digit > 0 || fractionalSeconds % d > 0) {
-          result.append((char) ('0' + digit));
-        }
-      }
-    }
-  }
+	private static final ZoneId ZULU = ZoneId.of("Z");
+
+	private static final Pattern PATTERN = Pattern.compile("(-?\\p{Digit}{4,})-(\\p{Digit}{2})-(\\p{Digit}{2})"
+			+ "T(\\p{Digit}{2}):(\\p{Digit}{2})(?::(\\p{Digit}{2})(\\.(\\p{Digit}{0,12}?)0*)?)?"
+			+ "(Z|([-+]\\p{Digit}{2}:\\p{Digit}{2}))?");
+
+	private static final EdmDateTimeOffset INSTANCE = new EdmDateTimeOffset();
+
+	public static EdmDateTimeOffset getInstance() {
+		return INSTANCE;
+	}
+
+	@Override
+	public Class<?> getDefaultType() {
+		return Timestamp.class;
+	}
+
+	@Override
+	protected <T> T internalValueOfString(final String value, final Boolean isNullable, final Integer maxLength,
+			final Integer precision, final Integer scale, final Boolean isUnicode, final Class<T> returnType)
+			throws EdmPrimitiveTypeException {
+		try {
+			ZonedDateTime zdt = parseZonedDateTime(value);
+
+			return convertZonedDateTime(zdt, returnType);
+		} catch (DateTimeParseException ex) {
+			throw new EdmPrimitiveTypeException("The literal '" + value + "' has illegal content.", ex);
+		} catch (final ClassCastException e) {
+			throw new EdmPrimitiveTypeException("The value type " + returnType + " is not supported.", e);
+		}
+	}
+
+	private static ZonedDateTime parseZonedDateTime(final String value) {
+		ZonedDateTime zdt;
+		try {
+			// ISO-8601 conform pattern
+			zdt = ZonedDateTime.parse(value);
+		} catch (DateTimeParseException ex) {
+			// for backward compatibility - allow patterns that don't specify a time zone
+			final Matcher matcher = PATTERN.matcher(value);
+			if (matcher.matches() && matcher.group(9) == null) {
+				zdt = ZonedDateTime.parse(value + "Z");
+			} else {
+				throw ex;
+			}
+		}
+		return zdt;
+	}
+
+	@SuppressWarnings("unchecked")
+	private static <T> T convertZonedDateTime(ZonedDateTime zdt, Class<T> returnType) {
+		if (returnType == ZonedDateTime.class) {
+			return (T) zdt;
+		} else if (returnType == Instant.class) {
+			return (T) zdt.toInstant();
+		} else if (returnType.isAssignableFrom(java.util.Date.class)) {
+			return (T) java.util.Date.from(zdt.toInstant());
+		} else if (returnType.isAssignableFrom(java.sql.Time.class)) {
+			return (T) new java.sql.Time(zdt.toInstant().truncatedTo(ChronoUnit.SECONDS).toEpochMilli());
+		} else if (returnType.isAssignableFrom(Timestamp.class)) {
+			return (T) Timestamp.from(zdt.toInstant());
+		} else if (returnType.isAssignableFrom(Long.class)) {
+			return (T) Long.valueOf(zdt.toInstant().toEpochMilli());
+		} else if (returnType.isAssignableFrom(Calendar.class)) {
+			return (T) GregorianCalendar.from(zdt);
+		} else {
+			throw new ClassCastException("unsupported return type " + returnType.getSimpleName());
```

Comment:
What if the return type is java.sql.Date?

Reply:
This not nice, as a date time offset semantically is an instant/timestamp whereas a java.sql.Date represents a date (without time components). But we can still support it for backwards compatibility. 

The interpretation is then that the millis stored in java.sql.Date represent the date-time-offset/instant/timestamp.